README.md
# k-metar-taf

[![Latest Release](https://img.shields.io/github/v/tag/kalisio/k-metar-taf?sort=semver&label=latest)](https://github.com/kalisio/k-metar-taf/releases)
[![Build Status](https://app.travis-ci.com/kalisio/k-metar-taf.svg?branch=master)](https://app.travis-ci.com/kalisio/k-metar-taf)

A [Krawler](https://kalisio.github.io/krawler/) based service to download [METAR](https://en.wikipedia.org/wiki/METAR) and [TAF](https://en.wikipedia.org/wiki/Terminal_aerodrome_forecast) data from the [Aviation Weather Center](https://www.aviationweather.gov/).

## Description

The **k-metar-taf** job allow to scrape **METAR** and **TAF** data using the [web services](https://www.aviationweather.gov/help/webservice) provided by the **Avioation Weather Center**. 

The downloaded data are stored within a [MongoDB](https://www.mongodb.com/) database and more precisely in 3 collections:
* `metar-taf-observations` that stores the **METAR** data
* `metar-taf-forecasts` that stores the **TAF** data
* `metar-taf-stations` that stores the stations data

All records are stored in [GeoJson](https://fr.wikipedia.org/wiki/GeoJSON) format.

The job is executed according a specific cron expression. By default, every hours.

## Implementation

Because webservices return a limited amount of data based on the query bounding box, the job consists of a set of tasks which query the webservice according a given bounding box.
The following illustrates the different bounding boxes used to defined the set of tasks to be executed:

![bboxes](./bboxes.jpg)

The subdivision is irregular in order to define smaller bounding boxes on areas of large amount of aerodromes (.e.g. north america and western europe).

## Configuration

| Variable | Description |
|--- | --- |
| `DB_URL` | The mongoDB database URL. The default value is `mongodb://127.0.0.1:27017/metar-taf` |
| `TTL` | The data time to live. It must be expressed in seconds and the default value is `604 800` (7 days) |
| `DATA` | The data to be scrapped. It must be either `metar` or `taf`. The default value is `metar` |
| `FOOTPRINT` | Writes the footprint of the different queries as a **GeoJSON** file. The default value is `false`. |
| `DEBUG` | Enables debug output. Set it to `krawler*` to enable full output. By default it is undefined. |

## Deployment

We personally use [Kargo](https://kalisio.github.io/kargo/) to deploy the service.

## Contributing

Please refer to [contribution section](./CONTRIBUTING.md) for more details.

## Authors

This project is sponsored by 

![Kalisio](https://s3.eu-central-1.amazonaws.com/kalisioscope/kalisio/kalisio-logo-black-256x84.png)

## License

This project is licensed under the MIT License - see the [license file](./LICENSE) for details


CHANGELOG.md
Please refer to GitHub issues for changelog starting from v0.3: https://github.com/kalisio/k-metar-taf/issues.

### 0.2.0 (2023-03-09)

##### New Features

*  Upgrade NodeJS to v16 and krawler to latest version (close [#3](https://github.com/kalisio/k-metar-taf/pull/3)) ([37d2c3ca](https://github.com/kalisio/k-metar-taf/commit/37d2c3ca222e01b1dc3414c2f29c8b63bab10b99))
*  compute composite name using the site and the icao id ([86b5c78b](https://github.com/kalisio/k-metar-taf/commit/86b5c78b94a6073f4f2d6056c434d729e24b60b5))

##### Bug Fixes

*  duplicate key in when running taf job ([dd30cda8](https://github.com/kalisio/k-metar-taf/commit/dd30cda8aa2356304a90207a137e42f0befe0005))

##### Other Changes

*  fix output path in fs store ([9403de7e](https://github.com/kalisio/k-metar-taf/commit/9403de7e9df06fc8141a5cba9dff064991653523))
*  Upgrade NodeJS to v16 and krawler to latest version [#3](https://github.com/kalisio/k-metar-taf/pull/3) ([4b7c399c](https://github.com/kalisio/k-metar-taf/commit/4b7c399c1e7f35af223973f38f7b9e22dd850750))
*  try on enhance performance when viewing ([029f0c0b](https://github.com/kalisio/k-metar-taf/commit/029f0c0b5346f78a7d3a8f608b546dcd6c7c3a34))
*  Duplicate data when scrapping TAF [#2](https://github.com/kalisio/k-metar-taf/pull/2) ([71105a80](https://github.com/kalisio/k-metar-taf/commit/71105a807a5b93c22f29907b17cdcc8d4e065051))
*  Duplicate data when scrapping TAF [#2](https://github.com/kalisio/k-metar-taf/pull/2) ([b90b6a8b](https://github.com/kalisio/k-metar-taf/commit/b90b6a8b21cfbe67561782d45723745cc13c997c))
*  Duplicate data when scrapping TAF [#2](https://github.com/kalisio/k-metar-taf/pull/2) ([968aeb68](https://github.com/kalisio/k-metar-taf/commit/968aeb685b6892f91384f9185d023a0284cab878))
*  Duplicate data when scrapping TAF ([28659b04](https://github.com/kalisio/k-metar-taf/commit/28659b04b09497f8235140ddf77f73098ef80dd7))
*  time reference for TAF ([d7bc16c1](https://github.com/kalisio/k-metar-taf/commit/d7bc16c1c50df9e59f0d49d5cb3885eaa0ef3af4))
*  Job is running out of memory (closed [#1](https://github.com/kalisio/k-metar-taf/pull/1)) ([a2b29e00](https://github.com/kalisio/k-metar-taf/commit/a2b29e0010e01fb67172201d0d8e4441828b4030))
* //github.com/kalisio/k-metar-taf ([a87c0353](https://github.com/kalisio/k-metar-taf/commit/a87c0353f9840013e1753f5bd33b45be27188557))

## 1.0.0 (2022-08-30)

##### New Features

*  Upgrade NodeJS to v16 and krawler to latest version (close [#3](https://github.com/kalisio/k-metar-taf/pull/3)) ([37d2c3ca](https://github.com/kalisio/k-metar-taf/commit/37d2c3ca222e01b1dc3414c2f29c8b63bab10b99))
*  compute composite name using the site and the icao id ([86b5c78b](https://github.com/kalisio/k-metar-taf/commit/86b5c78b94a6073f4f2d6056c434d729e24b60b5))

##### Bug Fixes

*  duplicate key in when running taf job ([dd30cda8](https://github.com/kalisio/k-metar-taf/commit/dd30cda8aa2356304a90207a137e42f0befe0005))

##### Other Changes

*  fix output path in fs store ([9403de7e](https://github.com/kalisio/k-metar-taf/commit/9403de7e9df06fc8141a5cba9dff064991653523))
*  Upgrade NodeJS to v16 and krawler to latest version [#3](https://github.com/kalisio/k-metar-taf/pull/3) ([4b7c399c](https://github.com/kalisio/k-metar-taf/commit/4b7c399c1e7f35af223973f38f7b9e22dd850750))
*  try on enhance performance when viewing ([029f0c0b](https://github.com/kalisio/k-metar-taf/commit/029f0c0b5346f78a7d3a8f608b546dcd6c7c3a34))
*  Duplicate data when scrapping TAF [#2](https://github.com/kalisio/k-metar-taf/pull/2) ([71105a80](https://github.com/kalisio/k-metar-taf/commit/71105a807a5b93c22f29907b17cdcc8d4e065051))
*  Duplicate data when scrapping TAF [#2](https://github.com/kalisio/k-metar-taf/pull/2) ([b90b6a8b](https://github.com/kalisio/k-metar-taf/commit/b90b6a8b21cfbe67561782d45723745cc13c997c))
*  Duplicate data when scrapping TAF [#2](https://github.com/kalisio/k-metar-taf/pull/2) ([968aeb68](https://github.com/kalisio/k-metar-taf/commit/968aeb685b6892f91384f9185d023a0284cab878))
*  Duplicate data when scrapping TAF ([28659b04](https://github.com/kalisio/k-metar-taf/commit/28659b04b09497f8235140ddf77f73098ef80dd7))
*  time reference for TAF ([d7bc16c1](https://github.com/kalisio/k-metar-taf/commit/d7bc16c1c50df9e59f0d49d5cb3885eaa0ef3af4))
*  Job is running out of memory (closed [#1](https://github.com/kalisio/k-metar-taf/pull/1)) ([a2b29e00](https://github.com/kalisio/k-metar-taf/commit/a2b29e0010e01fb67172201d0d8e4441828b4030))
* //github.com/kalisio/k-metar-taf ([a87c0353](https://github.com/kalisio/k-metar-taf/commit/a87c0353f9840013e1753f5bd33b45be27188557))








jobfile.js
import _ from 'lodash'
import { hooks } from '@kalisio/krawler'

const outputDir = './output'

const dbUrl = process.env.DB_URL || 'mongodb://127.0.0.1:27017/metar-taf'
const ttl = +process.env.TTL || (7 * 24 * 60 * 60)  // duration in seconds
const data = process.env.DATA || 'metar'
const footprint = process.env.FOOTPRINT ? process.env.FOOTPRINT === 'true' : false

// computed var
const script = _.capitalize(data) + 'JSON.php'
const collection = data === 'metar' ? 'observations' : 'forecasts'
const timePath = data === 'metar' ? 'properties.obsTime' : 'properties.validTimeFrom'

// Create a custom hook to generate tasks
let generateTasks = (options) => {
  return (hook) => {
    let footprintCollection = {
      type: 'FeatureCollection',
      'features': [] 
    }
    let tasks = []
    _.forEach(options.grids, grid => {
      const cellWidth = (grid.bbox[2] - grid.bbox[0]) / grid.columns
      const cellHeight = (grid.bbox[3] - grid.bbox[1]) / grid.rows
      for (let i = 0; i < grid.columns; i++) {
        for (let j = 0; j < grid.rows; j++) {
          const east=grid.bbox[0] + (i * cellWidth)
          const south=grid.bbox[1] + (j * cellHeight)
          const west=east + cellWidth
          const north=south + cellHeight
          footprintCollection.features.push({
            type: 'Feature',
            properties: {},
            geometry: {
              type: 'Polygon',
              coordinates: [
                [ [east, south], [west, south], [west, north], [east, north], [east, south] ]
              ]
            } 
          })
          console.log(`Task created for the bbox=${east},${south},${west},${north}`)
          let task = {
            options: {
              url: `https://www.aviationweather.gov/cgi-bin/json/${script}?priority=8&bbox=${east},${south},${west},${north}`
            }
          }
          tasks.push(task)
        }
      }
    })
    hook.data.tasks = tasks
    hook.data.footprint = footprintCollection
    return hook
  }
}
hooks.registerHook('generateTasks', generateTasks)

export default {
  id: 'metar-taf',
  store: 'memory',
  options: {
    workersLimit: 1
  },
  taskTemplate: {
    id: 'metar-taf/<%= taskId %>',
    type: 'http'
  },
  hooks: {
    tasks: {
      after: {
        readJson: {},
        transformJson: {
          transformPath: 'features',
          filter: { id: { $exists: true } } // Skips the first element
        },
        apply: {
          function: (item) => {
            console.log(`${item.options.url}: found ${item.data.features.length} data`)
            _.forEach(item.data.features, (feature) => {
              const featureTime = new Date(_.get(feature, timePath)).getTime()
              feature.properties.key = `${feature.id}-${featureTime}`
              feature.properties.name = `${feature.properties.site} [${feature.properties.id}]`
            })
          }
        },
        updateData: {
          hook: 'updateMongoCollection',
          collection: 'metar-taf-' + collection,
          filter: { 'properties.key': '<%= properties.key %>' },
          upsert : true,
          transform: {
            mapping: {
              [timePath]: { path: 'time', delete: false },
              'properties.temp': 'properties.temperature',
              'properties.dewp': 'properties.dewpoint',
              'properties.wspd': 'properties.windSpeed',
              'properties.wdir': 'properties.windDirection',
              'properties.wgst': 'properties.windGust',
              'properties.cover': 'properties.cloudCover',
              'properties.visib': 'properties.visibility',
              'properties.id': 'properties.icao'
            },
            omit: [
              'id',
              'properties.data',
              'properties.site'
            ],
            unitMapping: { time: { asDate: 'utc' } } 
          },
          chunkSize: 256
        },
        updateStations: {
          hook: 'updateMongoCollection',
          collection: 'metar-taf-stations',
          filter: { 'properties.icao': '<%= properties.icao %>' },
          upsert : true,
          transform: {
            omit: [ 
              'time',
              'properties.dataId',
              'properties.obsTime',
              'properties.issueTime',
              'properties.validTimeFrom',
              'properties.validTimeTo',
              'properties.validTime',
              'properties.timeGroup',
              'properties.fcstType',              
              'properties.temperature',
              'properties.dewpoint', 
              'properties.windSpeed', 
              'properties.windDirection',
              'properties.windGust',
              'properties.ceil',
              'properties.cloudCover',
              'properties.cldType1',
              'properties.cldCvg1',
              'properties.cldBas1',
              'properties.cldType2',
              'properties.cldCvg2',
              'properties.cldBas2',
              'properties.cldType3',
              'properties.cldCvg3',
              'properties.cldBas3',
              'properties.cldType4',
              'properties.cldCvg4',
              'properties.cldBas4',
              'properties.visibility',
              'properties.fltcat',
              'properties.altim',
              'properties.slp',
              'properties.wx',
              'properties.rawOb',
              'properties.rawTAF' ]
          },
          chunkSize: 256
        },
        clearData: {}
      }
    },
    jobs: {
      before: {
        createStores: [
          { id: 'memory' },
          { 
            id: 'fs', 
            options: { 
              path: outputDir
            } 
          }
        ],
        connectMongo: {
          url: dbUrl,
          // Required so that client is forwarded from job to tasks
          clientPath: 'taskTemplate.client'
        },
        createDataCollection: {
          hook: 'createMongoCollection',
          clientPath: 'taskTemplate.client',
          collection: 'metar-taf-' + collection,
          indices: [
            [{'properties.key': 1 }, { unique: true }],
            { 'properties.icao': 1 },
            [{ 'properties.icao': 1, time: -1 }, { background: true }],
            [{ 'properties.icao': 1, 'properties.temperature': 1, time: -1 }, { background: true }],
            [{ 'properties.icao': 1, 'properties.dewpoint': 1, time: -1 }, { background: true }],
            [{ 'properties.icao': 1, 'properties.windSpeed': 1, time: -1 }, { background: true }],
            [{ 'properties.icao': 1, 'properties.windDirection': 1, time: -1 }, { background: true }],
            [{ 'properties.icao': 1, 'properties.windGust': 1, time: -1 }, { background: true }],
            [{ 'properties.icao': 1, 'properties.cloudCover': 1, time: -1 }, { background: true }],
            [{ 'properties.icao': 1, 'properties.visibility': 1, time: -1 }, { background: true }],
            { 'properties.temperature': 1 },
            { 'properties.dewpoint': 1 },
            { 'properties.windSpeed': 1 },
            { 'properties.windDirection': 1 },
            { 'properties.windGust': 1 },
            { 'properties.cloudCover': 1 },
            { 'properties.visibility': 1 },
            [{ time: 1 }, { expireAfterSeconds: ttl }], // days in s
            { geometry: '2dsphere' }                                                                                                              
          ],
        },
        createStationsCollection: {
          hook: 'createMongoCollection',
          clientPath: 'taskTemplate.client',
          collection: 'metar-taf-stations',
          indices: [
            [{ 'properties.icao': 1 }, { unique: true }],
            { geometry: '2dsphere' }                                                                                                              
          ]
        },
        generateTasks: {
          grids: [
            { bbox: [-180, -80, 180, -40], columns: 2, rows: 1 },
            { bbox: [-180, -40, 180, 20], columns: 6, rows: 3 },
            { bbox: [-180, 20, -130, 60], columns: 1, rows: 2 },
            { bbox: [-130, 20, -60, 30], columns: 4, rows: 1 },
            { bbox: [-130, 30, -60, 50], columns: 8, rows: 4 },
            { bbox: [-130, 50, -60, 60], columns: 4, rows: 1 },
            { bbox: [-60, 20, 0, 60], columns: 1, rows: 2 },
            { bbox: [0, 20, 60, 40], columns: 4, rows: 2 },
            { bbox: [0, 40, 30, 60], columns: 4, rows: 4 },
            { bbox: [30, 40, 60, 60], columns: 2, rows: 2 },
            { bbox: [60, 20, 90, 60], columns: 1, rows: 4 },
            { bbox: [90, 20, 180, 60], columns: 3, rows: 2 },
            { bbox: [-180, 60, 180, 90], columns: 2, rows: 1 }
          ]
        }
      },
      after: {
        writeJson: {
          match: { predicate: () => footprint },
          dataPath: 'data.footprint',
          store: 'fs'
        },
        clearData: {},
        disconnectMongo: {
          clientPath: 'taskTemplate.client'
        },
        removeStores: ['memory', 'fs']
      },
      error: {
        disconnectMongo: {
          clientPath: 'taskTemplate.client'
        },
        removeStores: ['memory', 'fs']
      }
    }
  }
}
